### ISSUE TITLE
Make JMH benchmarks executable from command line and IDE

### ISSUE THREAD

**Issue by jeenbroekstra** [eclipse/rdf4j#2431]
Currently, most jmh benchmarks in the project are not (easily) executable without the use of a special plugin, which (apparently) is only available in the IntelliJ IDE. 

To make the benchmarks truly cross-platform and more easily executable on both the command line and in any IDE _without_ using a specialized plugin, we need to make sure that:

1. every benchmark class has a main method so it can be executed easily from within an IDE;
2. every maven module that has benchmarks produces a shaded jar so benchmarks can be run directly from the command line (see https://mkyong.com/java/java-jmh-benchmark-tutorial/ for some examples on how to set this up)

**Comment by jeenbroekstra** [comment]
The shade plugin is not an option for us unfortunately, because it cannot easily be used to generated a jar for the test scope in a format that jmh can work with. I'm investigating the use of the assembly plugin to produce a benchmark jar.

**Comment by jeenbroekstra** [comment]
The assembly plugin is not an option either because of its inability to merge SPI registry files.

**Comment by jeenbroekstra** [comment]
To make this work for the project as a whole, we'll move benchmark tests to their own modules. Each module will mirror the core module for which it provides benchmarks, for example for the memory store we'll have a separate module `rdfj4-benchmark-sail-memory`. The entire tree will live in a new parent module `benchmarks`, alongside core, compliance, and testsuites.

This way, we can easily:

1. run benchmarks from the IDE directly;
2. use the shade plugin to produce benchmark jar artifacts for execution from the command line;
3. ensure that benchmark code is _not_ deployed as part of releases

**Comment by hmottestad** [comment]
I don't see a need to run the benchmarks from command line. I've sometimes wondered why no one has created a benchmark saas solution like GitHub CI that would check the performance of a PR automatically. 

Main method for running the benchmarks is a good idea though. Pity no one has made a plugin for eclipse for running jmh benchmarks.

**Comment by jeenbroekstra** [comment]
> I don't see a need to run the benchmarks from command line.

There's a couple of reasons, but an important is this: running benchmarks from within an IDE can distort benchmark results. Almost every JMH article I've come across talks about this. That may be less of a problem if you control your environment carefully, but it doesn't hurt to take the IDE out of the equation. 

More importantly even: at its heart our build process must be independent from the IDE. That means that we should be able to run every aspect of our build (including tests and benchmarks) without having to rely on specific IDEs or IDE plugins. Ideally everything is doable via maven, but for now I'll settle for a `java -jar` solution.

>  I've sometimes wondered why no one has created a benchmark saas solution like GitHub CI that would check the performance of a PR automatically.

I think such things exist, custom-made, but running benchmarks on a CI can only give meaningful results if you have full control over the resources available. Github CI (at least the free variant that we use) can't be used for benchmarking because AFAIK there's no guarantee that the host machine will provide the same resources two times in a row. 

>  Pity no one has made a plugin for eclipse for running jmh benchmarks.

Probably because there's no need for a special plugin if you just run it as an application. Almost every JMH article I've come across shows how to do this, so it's pretty much the accepted way, it seems.

**Comment by hmottestad** [comment]
Yeah. The recommended way to run jmh benchmarks is to produce a jar. I found this very time consuming and tedious. Results have always been representative and I've not had any issues with anything being off when running directly from IntelliJ. Those same recommendations also recommend running a bunch of forks and a lot of iterations, so that each benchmarks takes a few hours. My experience with using benchmarks to drive the development of the ShaclSail is that you usually get good results after 5-10 iterations, C2 tends to kick in around there somewhere. I've ended up going with 20 warm up iterations and 10 measured iterations, to be on the safe side. 

It's also super useful to be able to right click on a single benchmark to run it, instead of having to type out the class and method name on the command line. 

Also very nice to have the benchmarks in the test directory so it's fast to compile and easy to run. 

I'll often create a few test while developing a new feature, then make some benchmarks once everything works to start focusing on performance.

**Comment by jeenbroekstra** [comment]
> Yeah. The recommended way to run jmh benchmarks is to produce a jar. I found this very time consuming and tedious. Results have always been representative and I've not had any issues with anything being off when running directly from IntelliJ. Those same recommendations also recommend running a bunch of forks and a lot of iterations, so that each benchmarks takes a few hours. My experience with using benchmarks to drive the development of the ShaclSail is that you usually get good results after 5-10 iterations, C2 tends to kick in around there somewhere. I've ended up going with 20 warm up iterations and 10 measured iterations, to be on the safe side.

I am not advocating that you shouldn't use IntelliJ + plugin to run these tests - it is clearly getting you results. I just want to make sure that we don't have our code set up such that they can _only_ be used if you use IntelliJ + plugin. That is not acceptable.

You say it's tedious to have to produce the benchmark jar. Of course it's more work than just right-clicking and going 'Run', but I don't quite understand why it would be such a chore: compared to the time it takes to actually run the benchmark, it doesn't seem that big a penalty to me. And as for running the specific benchmark from the command line: it is just a matter of typing `java -jar target/benchmark.jar DataFileBenchmark.write` and you're off (and with repeated runs, the command line history is your friend). Anyway, as said: if you prefer using the IDE for this, that's understandable and fine. I certainly don't want to force you to use the command line every time. I just want the option available, if at all possible.

However, I am curious to understand how, in practice, you actually use these benchmarks in your dev cycle. What does your workflow look like? 

1. You run a benchmark, wait anywhere between 10-20 minutes for it to complete, get some numbers. Presumably you copy those numbers somewhere to later reference them (so you can see how it changes as you make code improvements)? Do you keep them in a spreadsheet, or do you keep a text editor open somewhere to copy these results? How do you keep track of changes over time?
2. You make changes to the code that you hope/suspect/know will improve performance. As an aside: just having benchmarks numbers doesn't tell you much about where the problem is - do you also use a profiler at some point to find bottlenecks?
3. You compile, then run the benchmark again, get new numbers. If the numbers significantly improve, you keep the change, if not, you keep digging I presume? 

How often do you normally iterate through this loop? Presumably running the benchmark after _every_ change gets tedious quickly, regardless of how easy it is to start the benchmark. 

> It's also super useful to be able to right click on a single benchmark to run it, instead of having to type out the class and method name on the command line.

As said, I'm not against you doing that. What I _am_ against is a setup that doesn't also work on the command line and/or in IDEs other than IntelliJ. 

I think you're currently the only developer using jmh benchmarks in a serious manner in this project - and while that is to a large extent because you have more experience with them and are perhaps more focused on performance, it's at least partly also because the current setup makes it hard to run any benchmark you created in any environment other than IntelliJ. And it's something we can, I feel, fix relatively easily by one or two minor changes to how we set up these benchmarks. 

> Also very nice to have the benchmarks in the test directory so it's fast to compile and easy to run.

Why would it be faster to compile and run if it's in the test directory as opposed to a separate benchmark module? If it's in a separate module, as long as you have your project dependencies set up correctly, it should be just as fast to compile and run it, surely? In Eclipse at least, the individual maven modules all have direct dependencies on each other, so that if (for example) I make changes to the native store code, any modules that depend on the native store  _immediately_ see that change and use the changed code. Is that not the same in IntelliJ?

**Comment by jeenbroekstra** [comment]
Btw I'll also look into using the  compliance  modules  for keeping the benchmarks,  instead of introducing  even more separate modules.

**Comment by jeenbroekstra** [comment]
Hm, I'm backing away from the separate modules idea for now. While it enables command line execution I am having a hard time justifying the way it blows up the number of modules we have in the project.

Let's start with small steps first: adding a main method to every benchmark test. The rest we can reexamine later.

**Comment by jeenbroekstra** [comment]
Hm. I'm not happy with the way separate benchmark modules blow up the size of the project as a whole. 

I'm taking another look at using the assembly plugin. If we could use it to produce the benchmark jar, it would be possible to do so and keep the benchmarks in the test directory. By default the assembly plugin has a major drawback however: it can't deal with properly merging META-INF/services files, which can cause all sort of issues with our parsers etc. It might be possible to work around that though.

**Comment by jeenbroekstra** [comment]
Hah. Turns out my info on the assembly plugin was out of date. Never versions _can_ deal with the META-INF/services file properly. This is nice, gives us best of both worlds! We can keep the benchmarks in the test directory, and just configure project to produce a benchmark fat jar to use from the command line. 

Adding a main method still a good idea though as it makes running directly from the IDE easier.

**Comment by hmottestad** [comment]
I use Java Flight Recorder to profile. I launch it from the args in the @Fork annotation. I've found that the jmh stack profiler gives a lot of false positives, so I dropped that one. I once spent a day optimizing some stuff it said was taking 70% of the time, got it down to taking only 5% of the time but the program didn't run any faster at all. 

I've also found that the built in profiler in IntelliJ is pretty good, but it doesn't work with jmh. And also only available in the commercial version. 

As for my general process. 

I usually start off with a good idea of what is slow and why. I'll copy paste the benchmark results into a text editor and then start rewriting and tweaking. Occasionally I'll get stumped, or want to figure out why something is slower than expected. Then I'll use (JFRA) java flight recorder and record one specific benchmark and look at the profiling. I find that JFR gives great results but it fairly time consuming. That's why I like the new profiler in IntelliJ, since it's super simple to run and the results are very visual with a flame graph. That's how I discovered a string regex replace that was taking way too much resources compared to just a .startsWith(...) and a .substring(...). But I wouldn't have found it with JFR since it wasn't taking a lot of resources overall, or compared with other operations...but taking too much for what it was doing.
